Rewrite Installing Puppet agent during Provisioning (#2015)
theforeman/foreman-documentation: guides/common/modules/proc_installing-puppet-agent-during-host-provisioning.adoc

For more information, see {ContentManagementDocURL}Managing_Activation_Keys_content-management[Managing Activation Keys] in _{ContentManagementDocTitle}_.
endif::[]
ifdef::katello,orcharhino[]
.Prerequisites
* You created a Product and repository for the upstream Puppet agent, such as `\https://yum.puppet.com` or `\https://apt.puppet.com`, and synchronized the repository to {Project}.
For more information, see {ContentManagementDocURL}Importing_Content_content-management[Importing Content] in _{ContentManagementDocTitle}_.
* You created an activation key that enables the Puppet agent repository for hosts.
For more information, see {ContentManagementDocURL}Managing_Activation_Keys_content-management[Managing Activation Keys] in _{ContentManagementDocTitle}_.
endif::[]

.Procedure
. Navigate to *Hosts* > *Provisioning Templates*.
. Select a provisioning template depending on your host provisioning method.
For more information, see {ProvisioningDocURL}types-of-provisioning-templates_provisioning[Types of Provisioning Templates] in _{ProvisioningDocTitle}_.
ifndef::katello,orcharhino,satellite[]

> . Ensure the `puppetlabs_repo` snippet is included as follows:

Reviewer: Open question: I see it's copying what was there before with `puppet_setup`, but I wonder if it's clear to users that the default provisioning templates we ship (kickstart, preseed, etc) already do this. That is actually the reason we tell users to set the host parameter `enable-puppetlabs-repo`: it ensures the snippet gets loaded. So it feels like a redundant instruction.

Author: `enable-puppetlabs-repo` ensures both the snippet gets loaded in the default templates AND is also used inside the snippet.
But if we assume that the user uses the default provisioning templates, then both snippet instructions are redundant. Perhaps we should remove the whole part about editing a template?

Reviewer 2: Do we want to remove the section or should we merge this and continue if needed in a follow-up? @Lennonka @ekohl

Author: I was also thinking about adding a note to each snippet step that it's already included in the shipped templates. I'll do that instead removing the steps.

+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
<%= snippet 'puppetlabs_repo' %>
----
endif::[]
. Ensure the `puppet_setup` snippet is included as follows:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
<%= snippet 'puppet_setup' %>
----
ifndef::katello,orcharhino,satellite[]
. Enable the Puppet agent from the official Puppet repository using one of the following host parameters in global parameters, a host group, or for a single host:

* Add a host parameter named `enable-puppetlabs-repo` for the latest stable Puppet release in the unversioned repository.
* Add a host parameter named `enable-official-puppet7-repo` for the Puppet 7 release repository.

+
Select the *boolean* type and set the value to `true`.
endif::[]
ifdef::katello,orcharhino,satellite[]
. Enable the Puppet agent using a host parameter in global parameters, a host group, or for a single host.
Add a host parameter named `enable-puppet7`, select the *boolean* type, and set the value to `true`.
endif::[]
ifdef::katello,orcharhino,satellite[]

> . Ensure your host has access to the Puppet agent from {ProjectServer} by using an appropriate activation key.

Reviewer: Perhaps be explicit in what? My mind first went to firewalls when you said access
```suggestion
. Ensure your host has access to the Puppet agent packages from {ProjectServer} by using an appropriate activation key.
```

[id="Installing_Puppet_Agent_during_Host_Provisioning_{context}"]
= Installing Puppet Agent during Host Provisioning

Use this procedure to install the Puppet agent during the host provisioning process.

> ifdef::orcharhino,satellite[]

Reviewer: @maximiliankolb can you verify if orcharhino ships a vendored Puppet in their client repositories? Or it is like Katello and relies on upstream repositories?

* You created an activation key that enables the repository for hosts.
For more information, see {ContentManagementDocURL}Managing_Activation_Keys_content-management[Managing Activation Keys] in _{ContentManagementDocTitle}_.
* You assigned the activation key to a host group.
endif::[]

.Procedure
. Navigate to *Hosts > Provisioning Templates*.
. Select a provisioning template depending on your host provisioning method.
For more information, see {ProvisioningDocURL}types-of-provisioning-templates_provisioning[Types of Provisioning Templates] in _{ProvisioningDocTitle}_.
ifndef::katello,orcharhino,satellite[]
. Ensure the `puppetlabs_repo` snippet is included as follows:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
<%= snippet 'puppetlabs_repo' %>
----
endif::[]
. Ensure the `puppet_setup` snippet is included as follows:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
<%= snippet 'puppet_setup' %>
----
ifndef::katello,orcharhino,satellite[]
. Enable the official Puppet repository using a host parameter in global parameters, a host group, or for a single host.
+

> Add a host parameter named `enable-puppetlabs-repo` for the latest Puppet 7 or `enable-puppetlabs-puppet6-repo` for Puppet 6, select the `boolean` type, and set the value to `true`.

Reviewer: We have 4 versions now:
* `enable-puppetlabs-repo`: unversioned repository, latest stable release (so will be updated to Puppet 8 when it comes out later this year)
* `enable-official-puppet7-repo`: Puppet 7, latest current stable now
* `enable-puppetlabs-puppet6-repo`: Puppet 6, going EOL soon
* `enable-puppetlabs-puppet5-repo`: Puppet 5, already EOL

Probably shouldn't document the EOL ones, but giving the full overview.

endif::[]
. Enable Puppet agent using a host parameter in global parameters, a host group, or for a single host.
+
ifndef::satellite[]

> Add a host parameter named `enable-puppet7` for Puppet 7 or `enable-puppet6` for Puppet 6, select the `boolean` type and set the value to `true`.

Reviewer: This is redundant if you already set the repo parameter, so I'd recommend to leave it out.

[id="Installing_Puppet_Agent_during_Host_Provisioning_{context}"]
= Installing Puppet Agent during Host Provisioning

Use this procedure to install the Puppet agent during the host provisioning process.

ifdef::satellite[]
.Prerequisites
* You enabled and synchronized the *{project-client-name}* repository to {Project}.
For more information, see {ContentManagementDocURL}Importing_Content_content-management[Importing Content] in _{ContentManagementDocTitle}_.
* You created an activation key that enables the *{project-client-name}* repository for hosts.
For more information, see {ContentManagementDocURL}Managing_Activation_Keys_content-management[Managing Activation Keys] in _{ContentManagementDocTitle}_.
* You assigned the activation key to a host group.
endif::[]
ifdef::katello,orcharhino[]
.Prerequisites

> * You created a Product and repository for the upstream Puppet platform, such as `\https://yum.puppet.com` or `\https://apt.puppet.com`, and synchronized the repository to {Project}.

Reviewer 3: ```suggestion
* You created a Product and repository for the upstream Puppet agent, such as `\https://yum.puppet.com` or `\https://apt.puppet.com`, and synchronized the repository to {Project}.
```

For more information, see {ContentManagementDocURL}Importing_Content_content-management[Importing Content] in _{ContentManagementDocTitle}_.

> * You created an activation key that enables the repository for hosts.

Reviewer 3: ```suggestion
* You created an activation key that enables the Puppet agent repository for hosts.
```

ifndef::katello,orcharhino,satellite[]
. Ensure the `puppetlabs_repo` snippet is included as follows:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
<%= snippet 'puppetlabs_repo' %>
----
endif::[]
. Ensure the `puppet_setup` snippet is included as follows:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
<%= snippet 'puppet_setup' %>
----
ifndef::katello,orcharhino,satellite[]
. Enable the Puppet agent from the official Puppet repository using one of the following host parameters in global parameters, a host group, or for a single host:

* Add a host parameter named `enable-puppetlabs-repo` for the latest stable Puppet release in the unversioned repository.
* Add a host parameter named `enable-official-puppet7-repo` for the Puppet 7 release repository.
* Add a host parameter named `enable-puppetlabs-puppet6-repo` for the Puppet 6 release repository.

+
Select the `boolean` type and set the value to `true`.
endif::[]
ifdef::katello,orcharhino,satellite[]
. Enable the Puppet agent using a host parameter in global parameters, a host group, or for a single host.
+
endif::[]
ifdef::katello,orcharhino[]

> Add a host parameter named `enable-puppet7` for Puppet 7 or `enable-puppet6` for Puppet 6, select the `boolean` type and set the value to `true`.

Reviewer 3: ```suggestion
Add a host parameter named `enable-puppet7` for Puppet 7 or `enable-puppet6` for Puppet 6, select the `boolean` type, and set the value to `true`.
```

endif::[]
ifdef::satellite[]

> Add a host parameter named `enable-puppet7` for Puppet 7, select the `boolean` type and set the value to `true`.

Reviewer 3: ```suggestion
Add a host parameter named `enable-puppet7` for Puppet 7, select the `boolean` type, and set the value to `true`.
```

endif::[]
ifdef::katello,orcharhino,satellite[]

> . When you create the host, assign the host group that contains the appropriate activation key.

Reviewer 3: ```suggestion
. Ensure your host has access to the Puppet agent from {ProjectServer} by using an appropriate activation key.
```

Author: During provisioning (Create Host), the user cannot assign an activation key directly, but only indirectly using a host group instead.

Reviewer 3: If I go to hosts > create hosts; create a host; select a host group, then I can still overwrite the AK as host parameter or add one.

Author: If you mean the `kt_activation_keys` parameter, it seems that it's only interpreted like this in the Global Registration template and only when defined on the hostgroup level.
Or what parameters exactly do you mean?

Author: Oh, it's in a snippet. Never mind!

----
<%= snippet 'puppet_setup' %>
----
ifndef::katello,orcharhino,satellite[]
. Enable the Puppet agent from the official Puppet repository using one of the following host parameters in global parameters, a host group, or for a single host:

* Add a host parameter named `enable-puppetlabs-repo` for the latest stable Puppet release in the unversioned repository.
* Add a host parameter named `enable-official-puppet7-repo` for the Puppet 7 release repository.
* Add a host parameter named `enable-puppetlabs-puppet6-repo` for the Puppet 6 release repository.

+
Select the `boolean` type and set the value to `true`.
endif::[]
ifdef::katello[]
. Enable Puppet agent using a host parameter in global parameters, a host group, or for a single host.
+
Add a host parameter named `enable-puppet7` for Puppet 7 or `enable-puppet6` for Puppet 6, select the `boolean` type and set the value to `true`.
endif::[]
ifdef::orcharhino[]
. Enable Puppet agent using a host parameter in global parameters, a host group, or for a single host.
+
Add a host parameter named `enable-puppet6` for Puppet 6, select the `boolean` type and set the value to `true`.
endif::[]
ifdef::satellite[]

> . Enable Puppet agent using a host parameter in global parameters, a host group, or for a single host.

Reviewer: This paragraph looks the same as Katelloand orcharhino. Can we group them?

<%= snippet 'puppet_setup' %>
----
ifndef::katello,orcharhino,satellite[]
. Enable the Puppet agent from the official Puppet repository using one of the following host parameters in global parameters, a host group, or for a single host:

* Add a host parameter named `enable-puppetlabs-repo` for the latest stable Puppet release in the unversioned repository.
* Add a host parameter named `enable-official-puppet7-repo` for the Puppet 7 release repository.
* Add a host parameter named `enable-puppetlabs-puppet6-repo` for the Puppet 6 release repository.

+
Select the `boolean` type and set the value to `true`.
endif::[]
ifdef::katello[]
. Enable Puppet agent using a host parameter in global parameters, a host group, or for a single host.
+
Add a host parameter named `enable-puppet7` for Puppet 7 or `enable-puppet6` for Puppet 6, select the `boolean` type and set the value to `true`.
ifdef::orcharhino[]
. Enable Puppet agent using a host parameter in global parameters, a host group, or for a single host.
+

> Add a host parameter named `enable-puppet6` for Puppet 6, select the `boolean` type and set the value to `true`.

Reviewer: Any reason to use Puppet 6 here? @maximiliankolb?

I'll stress that Puppet 6 is about to be end of life so I'd urge you to get to Puppet 7.

Author: Waiting with another push, until we clarify this.
